yellowbox/extras/logstash.py

```diff
@@ -1,30 +1,263 @@
-from typing import Optional
-
-from docker import DockerClient
+import json
+import logging
+import selectors
+import socket
+import threading
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union, cast, TypedDict
+from weakref import WeakMethod
 
 from yellowbox.retry import RetrySpec
-from yellowbox.containers import get_ports, create_and_pull
-from yellowbox.subclasses import SingleContainerService, RunMixin
+from yellowbox.subclasses import YellowService
+from yellowbox.utils import _docker_host_name
+
+__all__ = ['LogstashService']
+_logger = logging.getLogger(__name__)
+
+_LISTEN_BACKLOG = 5  # Awaiting connections backlog count
+_STOP_TIMEOUT = 5  # Timeout for stopping the service
+
+# Sent on internal connection to signal closing of background thread
+_CLOSE_SENTINEL = b"\0"
+
+
+class LogstashService(YellowService):
+    """Implements a fake logging service that closely resembles Logstash.
+
+    Accepts external connections, with logs received in the "json_lines" format.
+
+    Attributes:
+        records: A list of all records received. You are welcome to modify,
+        clear or iterate over this during runtime. New logs will be added in the
+        order they were received. Each record is a dict - for possible keys see
+        attribute docs.
+        port: Dynamic port external service should connect to. LogstashService
+        automatically binds a free port during initialization.
+        encoding: Encoding of the json lines received. Defaults to utf-8 per
+        specification.
+        delimiter: Delimiter splitting between json objects. Defaults to b'\n' per
+        specification.
+    """
+    port: int
+    records: List[Dict[str, Any]]
+    """
+    Records generated by Python's numerous Logstash packages have at least the
+    following keys in common:
+        * @timestamp: str - Log timestamp in ISO-8601 format. 
+        * @version: str - Logstash format version (always 1)
+        * message: str - Log message.
+        * host: str - Host sending the message.
+        * path: str - Path to the module writing the log.
+        * tags: List[str] - ?
+        * type: str - "Logstash"
+        * level: str - An all upper-case name of the log level
+        * logger_name: str - Logger name
+        * stack_info: Optional[str] - Formatted stacktrace if one exists.
+
+    More keys may be added by the specific software sending the logs. 
+    """
+
+    delimiter: bytes = b"\n"
+    encoding: str = "utf-8"
+
+    def __init__(self) -> None:
+        """Initialize the service."""
+        self.records = []
+
+        root = socket.socket()
+        root.bind(("0.0.0.0", 0))
+        self._root = root
+
+        _background = WeakMethod(self._background_thread)
+        self._thread = threading.Thread(target=lambda: _background()(),
```

Comment:
why not this?

```suggestion
        self._thread = threading.Thread(target=self._background_thread,
```

Reply:
Cause then you create a cyclic reference and it won't clean up well.

Reply:
cyclic references are a small price to pay if it means we don't have to deal with weakrefs to methods as attributes. also if it's just to avoid cyclic refs then why not `target=lambda: self._background_thread()`?

Reply:
There are no weakref attributes, you don't have to deal with anything. Your suggestion leaves a cyclic reference and you don't gain anything by removing this.

Reply:
I still think that cyclic refs are fine in this case